### PR TITLE
[SPARK-20786][SQL]Improve ceil and floor handle the value which is not expected

### PR DESCRIPTION
**Pull request by heary-cao**
## What changes were proposed in this pull request?

spark-sql>SELECT ceil(1234567890123456);
1234567890123456

spark-sql>SELECT ceil(12345678901234567);
12345678901234568

spark-sql>SELECT ceil(123456789012345678);
123456789012345680

when the length of the getText is greater than 16. long to double will be precision loss.

but mysql handle the value is ok.

mysql> SELECT ceil(1234567890123456);
+------------------------+
| ceil(1234567890123456) |
+------------------------+
|       1234567890123456 |
+------------------------+
1 row in set (0.00 sec)

mysql> SELECT ceil(12345678901234567);
+-------------------------+
| ceil(12345678901234567) |
+-------------------------+
|       12345678901234567 |
+-------------------------+
1 row in set (0.00 sec)

mysql> SELECT ceil(123456789012345678);
+--------------------------+
| ceil(123456789012345678) |
+--------------------------+
|       123456789012345678 |
+--------------------------+
1 row in set (0.00 sec)

## How was this patch tested?

Supplement the unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -1337,12 +1337,14 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
    * possible, either a BigDecimal, a Long or an Integer is returned.
+   * when the length of the getText is greater than 16. long to double will be precision loss
+   * e.g:12345678901234567 => 1.2345678901234568e16
    */
   override def visitIntegerLiteral(ctx: IntegerLiteralContext): Literal = withOrigin(ctx) {
     BigDecimal(ctx.getText) match {
       case v if v.isValidInt =>
         Literal(v.intValue())
-      case v if v.isValidLong =>
+      case v if v.isValidLong && v.toString().length < 17 =>
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I think the question is, why is it a valid long but being used as a double (?) elsewhere?
Also you can use methods on BigDecimal to query its precision and scale. toString isn't the right approach

**Comment by hvanhovell** [comment]
+1 on Sean's comment. This is **not** a parser issue. Can you just fix this by adding `LongType` to `Ceil.inputTypes`?

**Comment by heary-cao** [comment]
@hvanhovell  @srowen 
I have modify it again. and` floor`  is same problem.
review please.
thanks.

**Comment by gatorsmile** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala [hunk 2]
```diff
@@ -173,6 +173,14 @@ class MathFunctionsSuite extends QueryTest with SharedSQLContext {
...
       sql("SELECT ceiling(0), ceiling(1), ceiling(1.5)"),
       Row(0L, 1L, 2L))
+    
+    checkAnswer(
+      sql("SELECT ceil(1234567890123456), ceil(12345678901234567)"),
+      Row(1234567890123456L, 12345678901234567L))
+
+    checkAnswer(
+      sql("SELECT ceiling(1234567890123456), ceiling(12345678901234567)"),
+      Row(1234567890123456L, 12345678901234567L))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala]: Could yo move all these new tests to the end of `operators.sql`? 

You can run the following command to generate the result files. 
```
SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"
```

Comment (heary-cao) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala]: ok, add new tests to the end of operators.sql.
please review it again.
thanks.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala]: Could you revert the changes you made in this file `MathFunctionsSuite.scala`?

**Comment by gatorsmile** [comment]
LGTM except one comment.

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala (hunk 2 above)
Comment (heary-cao) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala]: OK,
please review it again.
thanks.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master. 

Could you submit another PR to backport it to 2.2?

**Comment by ueshin** [comment]
@heary-cao Hi, I found `Ceil` and `Floor` modified by this pr produce different results between interpret and codegen.
We also need to modify `doGenCode()` to handle `LongType` correctly.
Can you fix it at a follow-up pr?

**Comment by rxin** [comment]
hm guys please don’t use the end-to-end tests to test expression behavior. use unit tests which automatically tests code gen, interpreted, and different data types.

**Comment by heary-cao** [comment]
@ueshin ,
thank you for your suggestion. 
modify `doGenCode() `pr: #18103
